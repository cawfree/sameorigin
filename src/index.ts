export * from './proxy';
export * from './puppeteer';
export * from './sameorigin';
export * from './server';
