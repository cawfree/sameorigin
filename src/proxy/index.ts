export * from './proxyRequest';
export * from './proxyRequestBody';
