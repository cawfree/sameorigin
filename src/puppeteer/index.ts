export * from './createPageWithSniffedRequests';
export * from './createStealthBrowserContext';
