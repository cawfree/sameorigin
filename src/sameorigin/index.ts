import axios, {AxiosRequestConfig, AxiosResponse} from 'axios';

import {createServer} from '../server';

export type SameOriginRequestConfig =
  & AxiosRequestConfig
  & Required<
    Pick<AxiosRequestConfig, 'baseURL'>
  >;

export const sameorigin = async ({
  squatURL,
  port = 3000,
  ...extras
}: Parameters<typeof createServer>[0] & {
  readonly squatURL: string;
}) => {
  const opts = await createServer({port, ...extras});

  const {data: {id}} = await axios({
    url: `http://localhost:${port}/open`,
    method: 'post',
    data: {uri: squatURL},
  });

  return [
    <ResponseType>({
      url: maybeUrl,
      baseURL,
      ...extras
    }: SameOriginRequestConfig): Promise<AxiosResponse<ResponseType>> => axios({
      ...extras,
      url: `http://localhost:${
        port
      }/${
        id
      }?baseUrl=${
        encodeURIComponent(String(baseURL))
      }${
        maybeUrl?.length ? `&url=${encodeURIComponent(String(maybeUrl))}` : ''
      }`,
    }),
    opts,
  ] as const;
};

export type SameOrigin = Awaited<ReturnType<typeof sameorigin>>;
