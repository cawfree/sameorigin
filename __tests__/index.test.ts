import 'jest';

import {sameorigin, SameOriginRequestConfig} from '../src';

jest.setTimeout(60 * 1000);

const requestAndClose = async (squatURL: string, params: SameOriginRequestConfig) => {
  const [axios, {close}] = await sameorigin({
    squatURL,
  });

  const {data} = await axios(params);

  await close();

  return data;
};

describe('index.test.ts', () => {

  it('opensea:carousel', async () => {
    const data = await requestAndClose(
      'https://opensea.io',
      {
        baseURL: 'https://opensea.io/__api/graphql/',
        method: 'post',
        data: {
          id: 'CategoryCarouselQuery',
          query: 'query CategoryCarouselQuery {\n  categoriesV2 {\n    slug\n    imageUrl\n    name\n  }\n}\n',
          variables: {},
        },
      }
    );
    expect(data).toMatchSnapshot();
  });

  it('opensea:categories', async () => {
    const data = await requestAndClose(
      'https://opensea.io',
      {
        baseURL: 'https://opensea.io/__api/graphql/',
        method: 'post',
        headers: {
          'x-build-id': 'cabc82dc7526c67efa0a5d98ebe6fa3d6f5d9cd8',
          'x-app-id': 'opensea-web',
          'x-signed-query': '5adb01ca3055fd2902dac5828b0f40b66279604bb472c77709df8c369c62ebcc',
          'x-viewer-address': '0x9ce69a314330687f1fb1ad1d397a0bb55d5e1d22',
        },
        data: {
          id: "CategoryScrollerQuery",
          query: "query CategoryScrollerQuery(\n  $categories: [CategoryV2Slug!]!\n  $highQualityFilter: Boolean\n) {\n  trendingCollectionsByCategory(first: 7, categories: $categories, highQualityFilter: $highQualityFilter) {\n    edges {\n      node {\n        id\n        ...HomePageCollectionCard_data_3C7EJl\n      }\n    }\n  }\n}\n\nfragment HomePageCollectionCardFooter_data_3C7EJl on CollectionType {\n  windowCollectionStats(statsTimeWindow: ONE_DAY) {\n    floorPrice {\n      unit\n      symbol\n    }\n    volume {\n      unit\n      symbol\n    }\n  }\n}\n\nfragment HomePageCollectionCard_data_3C7EJl on CollectionType {\n  banner\n  name\n  verificationStatus\n  drop {\n    ...useDropState_data\n    id\n  }\n  ...HomePageCollectionCardFooter_data_3C7EJl\n  ...collection_url\n}\n\nfragment collection_url on CollectionType {\n  slug\n  isCategory\n}\n\nfragment useDropState_data on DropType {\n  stages {\n    startTime\n    endTime\n    id\n  }\n  chainData {\n    mintedItemCount\n    totalItemCount\n  }\n}\n",
          variables: {"categories":["photography"],"highQualityFilter":true},
        },
      }
    );
    expect(data).toMatchSnapshot();
  });

  it('blur', async () => {
    const data = await requestAndClose(
      'https://blur.io/airdrop',
      {
        baseURL: 'https://core-api.prod.blur.io/v1',
        url: '/prices',
        method: 'get',
      }
    );
    expect(data).toMatchSnapshot();
  });
});
