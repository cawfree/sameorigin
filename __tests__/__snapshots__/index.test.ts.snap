// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`index.test.ts blur 1`] = `
{
  "ethereum": {
    "usd": 1735.74,
  },
}
`;

exports[`index.test.ts opensea:carousel 1`] = `
{
  "data": {
    "categoriesV2": [
      {
        "imageUrl": "https://opensea.io/static/images/categories/maverick-art.png",
        "name": "Art",
        "slug": "art",
      },
      {
        "imageUrl": "https://opensea.io/static/images/categories/maverick-gaming.png",
        "name": "Gaming",
        "slug": "gaming",
      },
      {
        "imageUrl": "https://opensea.io/static/images/categories/maverick-membership.png",
        "name": "Memberships",
        "slug": "memberships",
      },
      {
        "imageUrl": "https://opensea.io/static/images/categories/maverick-pfps.png",
        "name": "PFPs",
        "slug": "pfps",
      },
      {
        "imageUrl": "https://opensea.io/static/images/categories/maverick-photography.png",
        "name": "Photography",
        "slug": "photography",
      },
    ],
  },
}
`;

exports[`index.test.ts opensea:categories 1`] = `
{
  "data": {
    "trendingCollectionsByCategory": {
      "edges": [
        {
          "node": {
            "banner": "https://i.seadn.io/gae/uDuiXQtKsZI9kpV7ke_yuDJSALBtPRXfiWlmO9oeqoeLE8SAjP0gIMUh0FKD18lxsuYL5dsV7QLlClwQja9UBsT8xO4VMNV5S6lx3A?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6NjcwNDM3Mw==",
            "isCategory": false,
            "name": "Editions x Guido",
            "slug": "editionsxguido",
            "verificationStatus": "VERIFIED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.0098",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "1.740038889889999885",
              },
            },
          },
        },
        {
          "node": {
            "banner": "https://i.seadn.io/gcs/files/78be1282a7520fe8534e276681ff2bd2.gif?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6NzE5NTI2MQ==",
            "isCategory": false,
            "name": "Withhearts Editions",
            "slug": "withheartseditions",
            "verificationStatus": "SAFELISTED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.0035",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "0.58084990000000003",
              },
            },
          },
        },
        {
          "node": {
            "banner": "https://i.seadn.io/gcs/files/7b3278c432571f435004bbaf7ecea88a.jpg?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6MTkxNjA5OTg=",
            "isCategory": false,
            "name": "Rach Stewart Limited Editions",
            "slug": "rach-stewart-limited-editions",
            "verificationStatus": "SAFELISTED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.25",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "0.130000000000000004",
              },
            },
          },
        },
        {
          "node": {
            "banner": "https://i.seadn.io/gcs/files/fc8b9fa805a8be0ce66d5883af45e4f0.png?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6MjI1ODU4Mzc=",
            "isCategory": false,
            "name": "HARAMBE: LEGENDS",
            "slug": "harambe-legends",
            "verificationStatus": "SAFELISTED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.099",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "0.100000000000000006",
              },
            },
          },
        },
        {
          "node": {
            "banner": "https://i.seadn.io/gae/M3_gmKaQb3rUDyfpq9tUgUv36wY-fjyy8Z7olTgA2JXxqYApySlJi16axg_XkMgmhRGS2N0DWw9YZoHLoMCwfb5ZHCwm7-eMGvPRwg?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6NDk3MDU2MA==",
            "isCategory": false,
            "name": "IreneDAO",
            "slug": "irenedao",
            "verificationStatus": "SAFELISTED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.0881",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "0.080000000000000002",
              },
            },
          },
        },
        {
          "node": {
            "banner": "https://i.seadn.io/gae/FT8RdW-T7bo3Tyg2ElUxS9BqmZypSw8Je_UMD9PcQBGUlnnsw6m5LPuXQZEBDZ_LrOHOdji8P54I-I00eSUh75dYBkN3puTp7s0F?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6MTI5NjI1ODU=",
            "isCategory": false,
            "name": "First Day Out by DrifterShoots",
            "slug": "firstdayout",
            "verificationStatus": "SAFELISTED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.021",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "0.064199999999999993",
              },
            },
          },
        },
        {
          "node": {
            "banner": "https://i.seadn.io/gcs/files/9d0eb28a5eaa906991118cf536f921d9.jpg?w=500&auto=format",
            "drop": null,
            "id": "Q29sbGVjdGlvblR5cGU6NzAwMDMzMA==",
            "isCategory": false,
            "name": "FREE MINT Editions X Tylersjourney",
            "slug": "editionsxtylersjourney",
            "verificationStatus": "SAFELISTED",
            "windowCollectionStats": {
              "floorPrice": {
                "symbol": "ETH",
                "unit": "0.001494",
              },
              "volume": {
                "symbol": "ETH",
                "unit": "0.035166999999999997",
              },
            },
          },
        },
      ],
    },
  },
}
`;
